Python/pystats.c

```diff
@@ -0,0 +1,816 @@
+{
+    dest->mutex_sleeps = src->mutex_sleeps;
+    dest->qsbr_polls = src->qsbr_polls;
+    dest->world_stops = src->world_stops;
+}
+
+static void
+merge_rare_event_stats(RareEventStats *dest, const RareEventStats *src)
+{
+    dest->set_class += src->set_class;
+    dest->set_bases += src->set_bases;
+    dest->set_eval_frame_func += src->set_eval_frame_func;
+    dest->builtin_dict += src->builtin_dict;
+    dest->func_modification += src->func_modification;
+    dest->watched_dict_modification += src->watched_dict_modification;
+    dest->watched_globals_modification += src->watched_globals_modification;
+}
+
+static void
+merge_gc_stats_array(GCStats *dest, const GCStats *src)
+{
+    for (int i = 0; i < NUM_GENERATIONS; i++) {
+        dest[i].collections += src[i].collections;
+        dest[i].object_visits += src[i].object_visits;
+        dest[i].objects_collected += src[i].objects_collected;
+        dest[i].objects_transitively_reachable += src[i].objects_transitively_reachable;
+        dest[i].objects_not_transitively_reachable += src[i].objects_not_transitively_reachable;
+    }
+}
+
+void
+stats_zero_thread(_PyThreadStateImpl *tstate)
+{
+    // Zero the thread local stat counters
+    memset(tstate->pystats_struct, 0, sizeof(PyStats));
+}
+
+// merge stats for a single thread into the global structure
+void
+stats_merge_thread(_PyThreadStateImpl *tstate)
+{
+    PyStats *src = tstate->pystats_struct;
+    PyStats *dest = ((PyThreadState *)tstate)->interp->pystats_struct;
+
+    if (src == NULL || dest == NULL) {
+        return;
+    }
+
+    // Merge each category of stats using the helper functions.
+    merge_opcode_stats_array(dest->opcode_stats, src->opcode_stats);
+    merge_call_stats(&dest->call_stats, &src->call_stats);
+    merge_object_stats(&dest->object_stats, &src->object_stats);
+    merge_optimization_stats(&dest->optimization_stats, &src->optimization_stats);
+    merge_ft_stats(&dest->ft_stats, &src->ft_stats);
+    merge_rare_event_stats(&dest->rare_event_stats, &src->rare_event_stats);
+    merge_gc_stats_array(dest->gc_stats, src->gc_stats);
+}
+#endif // Py_GIL_DISABLED
+
+// toggle stats collection on or off for all threads
+static int
+stats_toggle_on_off(PyThreadState *tstate, int on)
+{
+    bool changed = false;
+    PyInterpreterState *interp = tstate->interp;
+    STATS_LOCK(interp);
+    if (on && interp->pystats_struct == NULL) {
+        PyStats *s = PyMem_RawCalloc(1, sizeof(PyStats));
+        if (s == NULL) {
+            STATS_UNLOCK(interp);
+            return -1;
+        }
+        FT_ATOMIC_STORE_PTR_RELAXED(interp->pystats_struct, s);
+    }
+    if (tstate->interp->pystats_enabled != on) {
+        FT_ATOMIC_STORE_INT_RELAXED(tstate->interp->pystats_enabled, on);
+        changed = true;
+    }
+    STATS_UNLOCK(interp);
+    if (!changed) {
+        return 0;
+    }
+    _PyEval_StopTheWorld(interp);
+    _Py_FOR_EACH_TSTATE_UNLOCKED(interp, ts) {
+        if (!ts->_status.active) {
```

Comment:
Since this loop happens when the world is paused, all the other threads will have _active=0, why is this check needed?

Reply:
Yup, you are correct.  This "active" check shouldn't be here.  I fixed this and a couple other toggle on/off issues I found.  I also added a unit test since obviously I didn't test this toggle functionality well enough since it wasn't working in other threads.